misc/synchronization.md

```diff
@@ -0,0 +1,368 @@
+  task depends on the previous write task.
+- Successive tasks that only read values a tile can be executed concurrently,
+  however they all depends on the previous write task and the next write task depends on
+  the completion of all the read tasks.
+
+## Matrix and Matrix Tiles basics
+
+The Matrix object (`dlaf::matrix::Matrix`) is the object which manage the setup of
+the correct dependencies of the tasks which involve any of its tiles.
+
+Matrix tiles (`dlaf::matrix::Tile`) can be accessed using two Matrix methods.
+- `operator()` returns a `hpx::future` of a tile containing the requested tile,
+- `read()` returns a copy of a `hpx::shared_future` of a tile with constant elements representing
+  the tile (allows only read operations on tile elements).
+
+Successive calls to `read()` with the same tile index return a copy of the same `shared_future`
+if no `operator()` was invoked.
+
+Both the Matrix and the Tile objects are not thread safe. The future mechanism ensures that
+each tile is either accessed in read-write mode by only one task, or in read-only mode by more tasks.
+On the other hand the user has to guarantee that a Matrix object is used only by one task at a time.
+For asynchronous scheduling [matrix views](#matrix-views) has to be used.
+
+### Basic Examples
+
+(Note: this examples are used to illustrate how dependencies are determined,
+however they cannot be compiled since they are simplified to improve readability.
+(e.g. constructors and template parameters are omitted)).
+
+- Example 1: Write dependencies.
+```
+  Matrix m;
+
+  hpx::dataflow(Task1, m({0, 0})));
+  hpx::dataflow(Task2, m({0, 0})));  // Depends on Task1.
+  hpx::dataflow(Task3, m({0, 1})));  // Different tile. No dependency on Task1 or Task2.
+
+// Resulting dependency graph:
+// Task1 - Task2
+//
+// Task3
+```
+- Example 2: Read dependencies.
+```
+  Matrix m;
+
+  hpx::dataflow(Task1, m({0, 0})));
+  hpx::dataflow(Task2, m({0, 1})));  // Different tile.
+  hpx::dataflow(Task3, m.read({0, 0}), m({1, 1})));  // Depends on Task1.
+  hpx::dataflow(Task4, m.read({0, 0}), m({0, 1})));  // Depends on Task1 and Task2. No dependency on Task3 (both only read Tile {0, 0})
+  hpx::dataflow(Task5, m({0, 0})));  // Depends on Task3 and Task4.
+  hpx::dataflow(Task6, m.read({0, 0})));  // Depends on Task5.
+
+// Resulting dependency graph:
+// Task1 - Task3 - Task 5 - Task6
+//       \       /
+// Task2 - Task4
+```
+
+### Exception handling
+
+If a task throws an exception during its execution the future returned by the asynchronous call
+contains the given exception.  Moreover any tile that was supposed to be written by the Task will
+be invalidated and any successive call to `operator()` and `read()` will return a future containing
+an exception. Note that because of C++ limitations we cannot propagate the correct exception type and
+message to tiles. Therefore they will contain a generic exception.
+
+## Matrix Views
```

Comment:
_"A MatrixView can be seen as a "sub-DAG" tool in the global DAG of the Matrix. The MatrixView gives us the chance to take control of a particular asynchronous step of the Matrix, and we can continue to work on the Matrix considering that everything we will do after the MatrixView, will happen after the MatrixView will be set as `done` (or `doneWrite` if we just want to release the writeability).
This is very handy, because it allows us to schedule tasks inside the sub-DAG of the MatrixView, and it will be like inserting new dependencies in between two already existing ones."_